server/src/main/java/com/linecorp/centraldogma/server/internal/storage/PurgeSchedulingService.java

```diff
@@ -0,0 +1,214 @@
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableScheduledFuture;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import com.linecorp.centraldogma.common.Author;
+import com.linecorp.centraldogma.server.command.Command;
+import com.linecorp.centraldogma.server.command.CommandExecutor;
+import com.linecorp.centraldogma.server.metadata.MetadataService;
+import com.linecorp.centraldogma.server.storage.project.ProjectManager;
+
+/**
+ * A service class for purging projects and repositories that were removed before.
+ */
+public class PurgeSchedulingService {
+
+    private static final Logger logger = LoggerFactory.getLogger(PurgeSchedulingService.class);
+
+    /**
+     * How often to run the storage purge schedules. i.e. every minute.
+     */
+    private static final Duration TICK = Duration.ofMinutes(1);
+
+    private final ProjectManager projectManager;
+    private final ScheduledExecutorService purgeWorker;
+    private final long maxRemovedRepositoryAgeMillis;
+    private final StoragePurgingScheduler storagePurgingScheduler = new StoragePurgingScheduler();
+
+    public PurgeSchedulingService(ProjectManager projectManager, ScheduledExecutorService purgeWorker,
+                                  long maxRemovedRepositoryAgeMillis) {
+        this.projectManager = requireNonNull(projectManager, "projectManager");
+        this.purgeWorker = requireNonNull(purgeWorker, "purgeWorker");
+        checkArgument(maxRemovedRepositoryAgeMillis >= 0,
+                      "maxRemovedRepositoryAgeMillis: %s (expected: >= 0)", maxRemovedRepositoryAgeMillis);
+        this.maxRemovedRepositoryAgeMillis = maxRemovedRepositoryAgeMillis;
+    }
+
+    public void start(CommandExecutor commandExecutor,
+                      MetadataService metadataService) {
+        if (isDisabled()) {
+            return;
+        }
+        requireNonNull(commandExecutor, "commandExecutor");
+        requireNonNull(metadataService, "metadataService");
+        cleanPurgedFiles();
+        storagePurgingScheduler.start(() -> {
+            try {
+                schedulePurgeData(commandExecutor, metadataService);
+            } catch (Exception e) {
+                logger.warn("Unexpected purging service failure", e);
+            }
+        });
+    }
+
+    public boolean isStarted() {
+        return storagePurgingScheduler.isStarted();
+    }
+
+    public void stop() {
+        if (!isDisabled()) {
+            storagePurgingScheduler.stop();
+        }
+    }
+
+    private void cleanPurgedFiles() {
+        projectManager.purgeMarked();
+        projectManager.list().forEach((unused, project) -> project.repos().purgeMarked());
+    }
+
+    @VisibleForTesting
+    void schedulePurgeData(CommandExecutor commandExecutor,
```

Comment:
I think this is not scheduling but removing. How about renaming to just `purgeData` or `purgeProjectAndRepository`?

Reply:
I agree, `purgeProjectAndRepository` looks good for me. Thanks.